packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2017,6 +2022,18 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
         context: ErrorDescription('while calling onSubmitted for $action'),
       ));
     }
+
+    // If `shouldUnfocus` is true but in case the developer cancelled the focus
+    // change in onSubmitted by focusing this input field again, reset the
+    // soft keyboard. See https://github.com/flutter/flutter/issues/84240.
+    //
+    // `_restartConnectionIfNeeded` creates a new TextInputConnection to replace
+    // the current one. This on iOS switches to a new input view and on Android
+    // restarts the input method, and in both cases the soft keyboard will be
+    // reset.
+    if (shouldUnfocus) {
+      scheduleMicrotask(_restartConnectionIfNeeded);
```

Comment:
This looks scary, why we need to put this in a microtask? also do we need to debounce?

Reply:
This is to check if the text field is indeed unfocused after the focus tree is updated. Yeah sounds like a good idea to schedule the restart only once.

Reply:
will that be determine in the same frame? if so, we may want postframecall back instead

Reply:
Done.

Reply:
> will that be determine in the same frame? if so, we may want postframecall back instead


The focus tree is updated twice per frame so I suspect in some extreme cases doing that in a postframe callback will be too late.

The logic here is trying to restart the connection only if the text field's scheduled focus change (in response to some text input actions) was cancelled immediately by the developer (in which case the text field's focus change callbacks will not be invoked, and the `TextInputConnection` will stay connected if we don't restart).

Reply:
I thought the scheduleMicrotask will also be pushed at the end of the frame? or am I missing something?

Reply:
The queued microtasks are run twice (the idle phase and the midFrameMicrotasks phase) each frame if I understand it correctly: https://api.flutter.dev/flutter/scheduler/SchedulerPhase-class.html

Reply:
I see, so it will run before the next frame starts. SGTM